activities.el

```diff
@@ -801,6 +851,116 @@ activity's name is NAME."
+                                        ((= cnt 1) unit)
+                                        (t units)))
+                      (fn age (cdr spec)))))))
+    (fn age activities--age-spec)))
+
+(defun activities--oldest-age (activities)
+  "Return the age in seconds of the oldest activity in ACTIVITIES."
+  (cl-loop for (_name . act) in activities maximize
+	   (cl-loop
+	    for type in '(default last)
+ 	    for state = (cl-struct-slot-value 'activities-activity type act)
+	    if state
+	    for etc = (activities-activity-state-etc state)
+	    maximize (float-time (time-since (map-elt etc 'time))))))
+
+(defun activities--annotate (max-age oldest-possible)
+  "Return an activity annotation function.
+MAX-AGE is the maximum age of any activity in seconds.
+OLDEST-POSSIBLE is the oldest age in the `vc-annotate-color-map'."
+  (lambda (name)
+    (when-let ((activity (map-elt activities-activities name)))
+      (let (data)
+	(dolist (type '(last default))
+	  (when-let ((state (cl-struct-slot-value 'activities-activity type activity)))
+	    (let* ((time (map-elt (activities-activity-state-etc state) 'time))
+		   (buffers-and-files (activities--buffers-and-files state)))
+	      (setf (alist-get type data)
+		    (list (and time (float-time (time-since time))) buffers-and-files)))))
+	(pcase-let* ((`(,default-age ,default-buffers-and-files) (map-elt data 'default))
+		     (`(,last-age ,last-buffers-and-files) (map-elt data 'last)) ;possibly nil
+		     (age (if last-age (min last-age default-age) default-age))
+		     (buffers-and-files (if last-age
+					    last-buffers-and-files
+					  default-buffers-and-files))
+		     (num-buffers (length buffers-and-files))
+		     (num-files (seq-count #'stringp (mapcar #'cdr buffers-and-files)))
+		     (dirtyp (when last-buffers-and-files
+			       (activities--buffers-and-files-differ-p
+				last-buffers-and-files
+				default-buffers-and-files)))
+		     (annotation (format "%s%s buf%s %s file%s "
+					 (if (activities-activity-active-p activity)
+					     (propertize "@" 'face 'bold) " ")
+					 (propertize (format "%2d" num-buffers) 'face 'success)
+					 (if (= num-buffers 1) " " "s")
+					 (propertize (format "%2d" num-files) 'face 'warning)
+					 (if (= num-files 1) " " "s")))
+		     (age-color (or (cdr (vc-annotate-compcar
+					  (* (/ age max-age) oldest-possible)
+					  vc-annotate-color-map))
+				    vc-annotate-very-old-color))
+		     (age-annotation (propertize
+				      (format "%15s" (apply #'format "[%d %s]"
+							    (activities--age age)))
+				      'face `(:foreground ,age-color
+							  :background ,vc-annotate-background)))
+		     (dirty-annotation (if dirtyp (propertize "*" 'face 'bold) " ")))
+	  (concat (propertize " " 'display
+			      `(space :align-to (- right ,(+ 1 (length annotation)
+							     (length age-annotation)))))
+		  annotation age-annotation dirty-annotation))))))
+
+(defun activities-sort-by-active-age (names)
```

Comment:
If this function is only used in one place, I'd like to avoid defining it as a top-level symbol.

Reply:
There is a customization option `activities-sort-function` for which this function is the default value.   This allows the user to specifying another sort.